### PR TITLE
Do not kill LSP server process when exception occurs in doc change handler, and prevent out of range exception

### LINKED ISSUE #2665: Language server crashes for the following program
```dafny
// producer.dfy
function Foo(x: int): bool {
  x == 2
}
```

```dafny
// consumer.dfy
include "./producer.dfy"

method Bar(x: int) 
  ensures Foo(x) {

}
```

Logging shows: 

```
Unhandled exception. System.ArgumentOutOfRangeException: Offset and length were out of bounds for the array or count is greater than the number of elements from index to the end of the source collection. (Parameter 'count')
   at System.IO.FileStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at Microsoft.Dafny.LanguageServer.Language.DiagnosticErrorReporter.CreateDiagnosticRelatedInformationFor(IToken token, String message)+MoveNext()
   at Microsoft.Dafny.LanguageServer.Language.DiagnosticErrorReporter.CreateDiagnosticRelatedInformationFor(IToken token, String message)+MoveNext()
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
   at Microsoft.Dafny.LanguageServer.Language.DiagnosticErrorReporter.ReportBoogieError(ErrorInformation error)
   at Microsoft.Dafny.LanguageServer.Workspace.Compilation.GetDiagnosticsFromResult(Document document, VerificationResult result)
   at Microsoft.Dafny.LanguageServer.Workspace.Compilation.HandleStatusUpdate(DocumentAfterTranslation document, IImplementationTask implementationTask, IVerificationStatus boogieStatus)
   at Microsoft.Dafny.LanguageServer.Workspace.Compilation.<>c__DisplayClass32_0.<VerifyTask>b__0(IVerificationStatus update)
   at System.Reactive.AnonymousSafeObserver`1.OnNext(T value) in /_/Rx.NET/Source/src/System.Reactive/AnonymousSafeObserver.cs:line 43
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value) in /_/Rx.NET/Source/src/System.Reactive/Internal/Sink.cs:line 49
   at System.Reactive.ObserveOnObserverNew`1.DrainStep(ConcurrentQueue`1 q) in /_/Rx.NET/Source/src/System.Reactive/Internal/ScheduledObserver.cs:line 577
   at System.Reactive.ObserveOnObserverNew`1.DrainShortRunning(IScheduler recursiveScheduler) in /_/Rx.NET/Source/src/System.Reactive/Internal/ScheduledObserver.cs:line 507
   at System.Reactive.ObserveOnObserverNew`1.<>c.<.cctor>b__17_0(IScheduler scheduler, ObserveOnObserverNew`1 self) in /_/Rx.NET/Source/src/System.Reactive/Internal/ScheduledObserver.cs:line 497
   at System.Reactive.Concurrency.ScheduledItem`2.InvokeCore() in /_/Rx.NET/Source/src/System.Reactive/Concurrency/ScheduledItem.cs:line 208
   at System.Reactive.Concurrency.EventLoopScheduler.Run() in /_/Rx.NET/Source/src/System.Reactive/Concurrency/EventLoopScheduler.cs:line 356
   at System.Threading.Thread.StartHelper.Callback(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.Thread.StartCallback()
[Error - 3:24:24 PM] Connection to server got closed. Server will not be restarted.
[Error - 3:24:24 PM] Request textDocument/codeAction failed.
Error: Connection got disposed.
```

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Changes
- Do not kill LSP server process when exception occurs in status update handler
- Prevent out of range exception, fixes #2665

### Testing
Manual testing with VSCode

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -1,12 +1,14 @@
+
 # Upcoming
 
 - feat: Support for the `{:opaque}` attibute on `const` (https://github.com/dafny-lang/dafny/pull/2545)
 - feat: Support for plugin-based code actions on the IDE (https://github.com/dafny-lang/dafny/pull/2021)
 - fix: Fixed a crash when parsing `newtype` in the parser (https://github.com/dafny-lang/dafny/pull/2649)
 - fix: Added missing error reporting position on string prefix check (https://github.com/dafny-lang/dafny/pull/2652)
+- fix: Prevent LSP server from exiting when a crash occurs
+- fix: Fix bug where LSP server would not show diagnostics that referred to included files
```
Comment (MikaelMayer) [inline RELEASE_NOTES.md:9]:
```suggestion
- fix: Prevent LSP server from exiting when a crash occurs (https://github.com/dafny-lang/dafny/pull/2664)
- fix: Fix bug where LSP server would not show diagnostics that referred to included files (https://github.com/dafny-lang/dafny/pull/2664)
```

File: RELEASE_NOTES.md [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 - fix: Check all compiled expressions to be compilable (https://github.com/dafny-lang/dafny/pull/2641)
 - fix: Better error reporting on counter-examples if an option is not provided (https://github.com/dafny-lang/dafny/pull/2650)
```
Comment (MikaelMayer) [inline RELEASE_NOTES.md:11]:
```suggestion
- fix: Better error reporting on counter-examples if an option is not provided (https://github.com/dafny-lang/dafny/pull/2650)


```

File: Source/DafnyLanguageServer/Language/DiagnosticErrorReporter.cs [hunk 3]
```diff
@@ -95,25 +95,22 @@ public class DiagnosticErrorReporter : ErrorReporter {
...
-              file.Read(toRead, range.StartToken.pos, rangeLength);
-              file.Close();
-              var read = toRead.ToString();
-              if (read != null) {
-                message = FormatRelated(read);
-              } else {
-                message = "Related location (could not read file " + fileName + ")";
-              }
+              var content = File.ReadAllText(fileName);
+              message = FormatRelated(content.Substring(range.StartToken.pos, rangeLength));
```
Comment (MikaelMayer) [inline Source/DafnyLanguageServer/Language/DiagnosticErrorReporter.cs:106]: I guess trading the previously optimized code for this simple but costly one is worth the fact that we don't need to deal with encodings. That's fair.

**Review by MikaelMayer — COMMENTED** [review]
Just minor updates

File: Source/DafnyLanguageServer/Language/DiagnosticErrorReporter.cs (hunk 3 above)
Comment (keyboardDrummer) [reply Source/DafnyLanguageServer/Language/DiagnosticErrorReporter.cs:106]: I think all included files should be loaded into memory and available through some in memory filesystem, so then there wouldn't be any extra file read here.

**Review by RustanLeino — APPROVED** [review]

**Review by atomb — DISMISSED** [review]
This looks good to me. It has the further advantage that it'll properly respect the encoding of the file (assuming that the token position and length do) rather than using byte offsets.

**Review by fabiomadge — APPROVED** [review]

**Pull request merged**